### PR TITLE
use pyfaidx for FASTA sequences in Dataset and DataLoader

### PR DESCRIPTION
**Pull request by yangtcai**
hi, @williamstark01, I have used the pyfadix to generate the dataloader, however, there is a big problem, I test the dataloader performance on my local computer, and the generating speed is very slow.
image
the time labeled in the figure is just one chromosome will occupied.
Should we move this dynamic process into the data generation stage?

### CODE REVIEW

**Comment by williamstark01** [comment]
Hey Yantong, good job, you are almost there!

The idea with the Dataset / DataLoader distinction is to do only light computation during the Dataset object generation, and delay heavier computation to the batch generation stage. The heavier computation code would be moved to the `__getitem__` method, so it will be run by the batch generation worker processes during training. (And we can have many workers, so the training speed isn't affected.)

For this dataset, you would create the pyfaidx Fasta object(s) ("genome") during initialization and load the repeat annotation file(s) ("annotations"), but do no further processing.

Then, in `__getitem__` you would use the chromosome name and the coordinates as input, and generate and retrieve the sequence from the corresponding stored Fasta object using the repeat annotations.

It would be maybe something like this:
```
    def __init__(self, ...):
        self.genome = Fasta(genome_fasta_path)
        self.annotations = <load annotation CSV(s)>

    def __getitem__(self, sequence_coordinates: tuple):
        <get complete sequence from self.genome using sequence_coordinates>
        <get desired repeat annotations from self.annotations using the sequence_coordinates, for example all annotations for LTR>
        # Is anything else needed?
        return <sample data structure>
```

How does that sound?

A couple more suggestions:

You could split the original hg38 FASTA to multiple files, one per chromosome, so that we can load a subset of them during testing and experimentation. (Probably just one initially to be able to run quick experiments.) You would then load them in a dictionary of pyfaidx Fasta objects in the Dataset.

It's better to use pandas to load CSV files and use it for filtering entries, it's very fast and convenient for this task. (With a single command you can get all entries in a specific start - end range.)

**Review by williamstark01 — CHANGES_REQUESTED** [review]
Hey Yantong, good job, you are almost there!

The idea with the Dataset / DataLoader distinction is to do only light computation during the Dataset object generation, and delay heavier computation to the batch generation stage. The heavier computation code would be moved to the `__getitem__` method, so it will be run by the batch generation worker processes during training. (And we can have many workers, so the training speed isn't affected.)

For this dataset, you would create the pyfaidx Fasta object(s) ("genome") during initialization and load the repeat annotation file(s) ("annotations"), but do no further processing.

Then, in `__getitem__` you would use the chromosome name and the coordinates as input, and generate and retrieve the sequence from the corresponding stored Fasta object using the repeat annotations.

It would be maybe something like this:
```
    def __init__(self, ...):
        self.genome = Fasta(genome_fasta_path)
        self.annotations = <load annotation CSV(s)>

    def __getitem__(self, sequence_coordinates: tuple):
        <get complete sequence from self.genome using sequence_coordinates>
        <get desired repeat annotations from self.annotations using the sequence_coordinates, for example all annotations for LTR>
        # Is anything else needed?
        return <sample data structure>
```

How does that sound?

A couple more suggestions:

You could split the original hg38 FASTA to multiple files, one per chromosome, so that we can load a subset of them during testing and experimentation. (Probably just one initially to be able to run quick experiments.) You would then load them in a dictionary of pyfaidx Fasta objects in the Dataset.

It's better to use pandas to load CSV files and use it for filtering entries, it's very fast and convenient for this task. (With a single command you can get all entries in a specific start - end range.)

File: dataloader.py [hunk 1]
```diff
@@ -14,35 +17,48 @@ def __init__(
...
-        for i in range(1, len(seq_results)):
-            seq_arrays = np.concatenate((seq_arrays, seq_results[i]), axis=0)
-        return seq_arrays
+    def find_corresponding_label(self, seq_start: int, seq_end: int, labels):
+        results = []
+        for label in labels:
+            align_start, align_end = int(label[0]), int(label[1])
+            if align_start > align_end:
+                continue
+            if seq_start <= align_start and align_end <= seq_end:
```
Comment (yangtcai) [inline dataloader.py:44]: For now, I skip the situation about the information on the reverse strand, it seems like FASTA file only contains the information of forward strand, should we consider it?

Comment (EreboPSilva) [reply dataloader.py:44]: Hi Yantong. **Let me check the data you're using and I'll let you know.**
I would say that, for the time being, you can ignore it as this is still sort of a proof of concept (in the sense that there are more functionalities missing, such as different types of repeats). To extract the reverse complement from a fasta file is really simple, in any case.

Comment (EreboPSilva) [reply dataloader.py:44]: Hi again, just wanted to double check, but as suspected, **you should consider the reverse strand**.
As I said, this should be really easy to fix (a couple of one-line conditionals came to mind, even). Let me know if I can be of further assistance.
Once more, if you want to finish the proof of concept before, I guess that would be ok... But I don't see a reason for waiting.

Comment (yangtcai) [reply dataloader.py:44]: Thanks for your suggestion! It solves my confusion. I will add it to the dataloador.:D

**Comment by yangtcai** [comment]
When I implement the dataset, there is an issue that some subsequence has the corresponding labels, which means there is no LTR in that subsequence, should we delete this subsequence in the dataset?

**Comment by williamstark01** [comment]
> When I implement the dataset, there is an issue that some subsequence has the corresponding labels, which means there is no LTR in that subsequence, should we delete this subsequence in the dataset?

@yangtcai Subsequences that don't contain the desired repeat type could also be useful, being additional negative examples. For this reason, I'd say we could keep a number of them, but not all of them if they are too many. What do you think?

And great work with the new dataset and dataloader! When the PR is complete request a review again to take a closer look at the code.

**Review by williamstark01 — APPROVED** [review]
Really excellent work! It's great to see your attention to detail. I did some small edits while going through the code, I'm pushing them in a separate PR now.

A couple of comments:

As the reference genome is soft-masked, we need to convert all sequences to upper case, so the network doesn't see this information about repeats. Will you take a look at that?

Was that a hidden joke? :D
```
from turtle import forward
```

The subsequence length and overlap would best be converted to attributes of the network. (Simply save these values as attributes for now, we can change them to hyperparameters in the future.)
```
        genome_index = index * 1500
        # produce sequence with overlap 1500 of length 2000
        start = genome_index
        end = genome_index + 2000
```

**Comment by yangtcai** [comment]
Hi, @williamstark01, I added an extension in my vscode, and that extension is used for automatically completing the library, and it makes the joke of turtle😅. I think changing them into hyperparameters is a good idea, I will change it later.
